### PR TITLE
Reverse Cuthill-Mckee ordering algorithm 

### PR DESCRIPTION
**Pull request by abhinavmehndiratta**
#1123 
Although it does not cause any speed up in algorithms (since the cost of calculating the ordering is a lot) it could be an addition to LightGraphs as a feature.
Here are some benchmarks for reference -  
```julia
julia> using LightGraphs,BenchmarkTools

julia> g1 = SimpleGraph{UInt8}(200,100)
{200, 100} undirected simple UInt8 graph

julia> g2 = g1
{200, 100} undirected simple UInt8 graph

julia> @btime LightGraphs.Experimental.all_subgraphisomorph(g1, g2)
  207.946 μs (32 allocations: 28.11 KiB)
Channel{Array{Tuple{UInt8,UInt8},1}}(sz_max:0,sz_curr:1)

julia> @btime LightGraphs.Experimental.all_subgraphisomorph(g1[reverse_cuthill_mckee(g1)], g2[reverse_cuthill_mckee(g2)])
  677.395 μs (5350 allocations: 614.16 KiB)
Channel{Array{Tuple{UInt8,UInt8},1}}(sz_max:0,sz_curr:1)

julia> g3 = g1[reverse_cuthill_mckee(g1)]
{200, 100} undirected simple UInt8 graph

julia> g4 = g2[reverse_cuthill_mckee(g2)]
{200, 100} undirected simple UInt8 graph

julia> @btime LightGraphs.Experimental.all_subgraphisomorph(g3, g4)
  146.934 μs (32 allocations: 28.11 KiB)
Channel{Array{Tuple{UInt8,UInt8},1}}(sz_max:0,sz_curr:1)

```

```julia
julia> spy(adjacency_matrix(g1))
              Sparsity Pattern
       ┌──────────────────────────────┐    
     1 │⠀⡠⠌⠀⡀⠠⢀⠀⠀⠀⠀⠢⠀⠀⠀⠀⠠⠂⠠⠀⠀⠁⢀⠄⠀⢀⠀⠀⠀⠀│ > 0
       │⠂⠁⠀⠀⡀⠀⠀⠠⠀⢀⠀⠀⢀⠀⠀⡀⠀⠀⡁⠀⢁⠀⢂⡀⠁⠂⡀⠂⠀⠀│ < 0
       │⠀⡈⠀⠈⠀⠀⠀⠒⠀⢀⠂⠀⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⡁⠈⠀⠇⠀⠀│    
       │⠀⠐⠀⡀⢠⠀⠀⠀⠔⣠⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠠⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⢀⠀⢀⠐⣡⠀⠀⠈⠄⠀⠀⡐⠀⠀⠀⠀⠀⢄⠀⠀⢀⠀⠀⠀⠀⡀⠀│    
       │⠠⡀⠀⠀⠈⠀⠀⠀⠂⠄⠀⡠⠀⡈⡀⠀⢀⠀⠀⠀⡀⠠⠀⠁⠀⠀⠈⠠⢀⠀│    
       │⠀⠀⠀⠐⠁⠀⠀⠀⠀⠀⡀⠠⠀⠀⠀⢀⠄⠂⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠂⠀│    
       │⠀⠀⠀⠠⠀⠀⠀⠀⠐⠈⠀⠈⠀⢀⠀⠀⠄⠀⠐⠀⠀⠀⠀⠀⠀⢀⠐⠀⠀⢀│    
       │⠠⠂⠀⠀⠀⠀⠀⠀⠀⠀⠀⠐⠠⠁⠀⠁⠀⠀⠈⠀⠐⠐⠄⢀⠄⠀⠀⠁⠀⠈│    
       │⠀⠂⠁⠈⠀⠀⠀⠀⠀⠀⠀⠀⠁⠀⠐⠀⠂⠀⠊⡠⠀⠀⠀⠀⠀⠀⠀⠤⠀⠀│    
       │⠄⠀⠁⠐⠀⠀⠀⠂⠀⠑⠀⡈⠀⠀⠀⠀⢐⠀⠀⠀⠀⠀⠉⠀⠐⠄⠀⠀⠀⠀│    
       │⠀⠔⠈⠰⡀⠀⠀⠀⠀⢀⠄⠀⠀⠀⠀⠀⠀⢁⠀⠀⠃⠀⠀⠀⠀⠀⠁⠀⠀⠀│    
       │⠀⢀⠡⠀⡁⠈⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⠀⠁⠀⠀⠐⠄⠀⠀⠀⠀⠀⠈⠄⠀│    
       │⠀⠀⠠⠈⠤⠄⠀⠀⠀⠀⠂⡀⠀⠀⠐⠀⠄⠀⠀⡄⠀⠀⠁⠀⡀⠀⠀⠀⠀⡀│    
   200 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠀⠐⠈⠀⠀⢀⡀⠀⠀⠀⠀⠀⠀⠀⠀⠁⠀⠠⢀⠐│    
       └──────────────────────────────┘    
       1                            200
                  nz = 200

julia> spy(adjacency_matrix(g3))
              Sparsity Pattern
       ┌──────────────────────────────┐    
     1 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ > 0
       │⠀⠀⠁⢀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ < 0
       │⠀⠀⠀⠀⠚⡠⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠑⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠁⣤⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠊⢄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠱⢂⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⢛⡰⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠻⠂⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠛⠄⡄⠀⠀⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⢡⠒⢄⠀⠀⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠑⢎⠑⡄⠀⠀⠀│    
       │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠉⠤⣣⡀⠀│    
   200 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠺⣢│    
       └──────────────────────────────┘    
       1                            200
                  nz = 200
```

### CODE REVIEW

**Comment by simonschoelly** [comment]
Don't worry about the cost of ordering, the idea is, that one can use this algorithm for preprocessing a graph, so that it will run faster afterwards. Besides, it looks as if `induced_subgraph` is not implemented as efficient, as it could be. So you should calculate the permutations and then do the benchmarks afterwards.

I've realized that graph isomorphism might actually not be the best algorithm for testing the efficiency here,  the problem is, that the algorithm has a different running time, depending on the permutation, because that decides the order of the search tree. So if you would test with graph isomorphism, you would have to average over a lot of different graphs.
I've tested your algorithm with `connected_components` and `bfs_parents` and in some cases it performs up to 3 times faster than a random permutation.

The other thing is, that you are using an Erdös-Renyi graph for benchmarking. While that is usually ok, you should never only use this kind of graph for benchmarking. The reason is, that while such random graphs are from a mathematic view the most common graphs, they always have similar properties and do not not resemble graphs that can be found in real life. 
So what you want to use here for testing are graphs, that have better locality and can be clustered. Such graphs arise for example in social networks or from geometric problems. I had some success with using `Grid`, `dorogovtsev_mendes` and `euclidean_graph` - you might have to play around with the parameters.
For existing graphs you can use the package  `SNAPDatasets.jl` with the command `loadsnap(:facebook_combined)` to load a social network graph with roughly 4000 vertices. And if you have a lot of memory, you can use the package `MatrixDepot` with the command `matrixdepot("DIMACS10/road_usa")` to load the adjacency matrix of a huge road network with roughly 20 million vertices.

File: src/matrix_ordering.jl [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            end
+        end
+        sort!(u_nbrs, by = v -> degree(g, v))
+        for v in u_nbrs push!(q, v) end
+    end
+end
+
+function reverse_cuthill_mckee(g::SimpleGraph)
+    T = eltype(g)
+    cm = Vector{T}()
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: Maybe give a more descriptive name to `cm`. Such as `result`, `result_order`, `vertex_permutation` for example.

File: src/matrix_ordering.jl [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    visited = falses(nv(g))
+    for i in vertices(g)
+        if !visited[i]
+            connected_rcm(g, visited, cm, i)
+        end
+    end
+    return reverse(cm)
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: Isn't there also a forward cuthill_mckee, that does not use the reverse? Would it be better then, to implement the forward algorithm, or does this has some disadvantages?

File: src/matrix_ordering.jl [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                push!(u_nbrs, v)
+                visited[v] = true
+            end
+        end
+        sort!(u_nbrs, by = v -> degree(g, v))
+        for v in u_nbrs push!(q, v) end
+    end
+end
+
+function reverse_cuthill_mckee(g::SimpleGraph)
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: Maybe we would need a more descriptive name, so this method has better discoverability. Such as `rcm_vertex_permutation` or `reverse_cuthill_mckee_vertex_perm` if that is not too long.

File: src/matrix_ordering.jl [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        for v in outneighbors(g, u)
+            if degree(g, v) < degree(g, u)
+               u = v
+            end
+        end
+    end
+    return u
+end
+
+function connected_rcm(g::SimpleGraph, visited, cm, src)
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: This will modify `cm` and visited, so we should better name it `connected_rcm!`.

File: src/matrix_ordering.jl [hunk 5]
```diff
@@ -0,0 +1,49 @@
+function pseudo_peripheral_node(g::SimpleGraph, src)
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: Add some small comment, on what a what a "pseudo perpipheral node " is.

File: src/matrix_ordering.jl [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    T = eltype(g)
+    u::T = src
+    lp = 0
+    while true
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: Isn't there a problem here, when the graph contains disconnected components? I think in this case, `lp` will be set to `typemax(T)` in the first iteration and therefore the loop will stop in the second iteration.

File: src/matrix_ordering.jl [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    return u
+end
+
+function connected_rcm(g::SimpleGraph, visited, cm, src)
+    T = eltype(g)
+    start::T = pseudo_peripheral_node(g, src)
+    q = Vector{T}()
+    push!(q, start)
+    while !isempty(q)
+        u::T = popfirst!(q)
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: no need for `u::T` here, as `popfirst!(q)` should always return an element of type `T` and `u` will never be assigned any other value.

File: src/matrix_ordering.jl [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        visited[u] = true
+        push!(cm, u)
+        u_nbrs = Vector{T}()
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: Can you pull `u_nbrs = Vector{T}()` out of the loop and just use `empty!(u_nbrs)` in every iteration? You will probably save some allocations this way.

File: src/matrix_ordering.jl [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    end
+    return u
+end
+
+function connected_rcm(g::SimpleGraph, visited, cm, src)
+    T = eltype(g)
+    start::T = pseudo_peripheral_node(g, src)
+    q = Vector{T}()
+    push!(q, start)
+    while !isempty(q)
```
Comment (simonschoelly) [inline src/matrix_ordering.jl]: Use `@inbounds`

**Comment by simonschoelly** [comment]
Eventually, this function will also need tests.

File: src/matrix_ordering.jl (hunk 2 above)
Comment (abhinavmehndiratta) [reply src/matrix_ordering.jl]: rcm has better sparsity reduction than the forward one

File: src/matrix_ordering.jl (hunk 6 above)
Comment (abhinavmehndiratta) [reply src/matrix_ordering.jl]: yes i should change it to check only the reachable vertices (or make gdistance 0 where it returns typemax)

File: src/matrix_ordering.jl (hunk 2 above)
Comment (simonschoelly) [reply src/matrix_ordering.jl]: Not sure what that exactly means, but then we should probably leave it that way.
You could use `reverse!` which will modify `cm`, so you don't allocate additional memory for storing the vertices.

**Comment by sbromberger** [comment]
Couple of notes:

- On the assumption that there may be several graph ordering algorithms that people may want to use, perhaps it's better to make a submodule (or at least a subdirectory) for this?

- I'm not entirely convinced this sort of thing belongs in LG (as opposed to some more general matrix package) but I'm willing to be persuaded.

**Comment by simonschoelly** [comment]
The algorithm itself uses the features from LightGraphs and not from linear algebra and it can also speed up graph algorithms, so I still think it belongs here. But there might be other, better algorithms that we would have to implement, so that we could compare.

How about putting this into the Experimental submodule for now? Then we won't commit to having this function forever and are open to changes in the future.

**Comment by sbromberger** [comment]
I think it's ok outside of Experimental, but we should have it in a separate directory and possibly a separate submodule.

**Comment by abhinavmehndiratta** [comment]
@sbromberger so should I create a separate directory for this ?

**Comment by sbromberger** [comment]
> should I create a separate directory for this ?

If we plan on adding more ordering algorithms, then yes. Otherwise, this is fine for now.

Have you seen @simonschoelly's comments? Are they resolved yet?

**Comment by abhinavmehndiratta** [comment]
> If we plan on adding more ordering algorithms, then yes. Otherwise, this is fine for now.
> 
> Have you seen @simonschoelly's comments? Are they resolved yet?

I am a bit caught up in my mid semester exams the moment but I'll try to resolve the issues.

**Comment by simonschoelly** [comment]
In that case, write your mid term exams first - this here can wait.

**Comment by sbromberger** [comment]
Thinking about this - does it make sense to have a "Preprocessing" submodule?

**Comment by abhinavmehndiratta** [comment]
> Thinking about this - does it make sense to have a "Preprocessing" submodule?

Are there any other other algorithms which could be added if we create this module ? Otherwise, I think this should be fine for now.

**Comment by simonschoelly** [comment]
Btw. some stuff happening here: https://discourse.julialang.org/t/implementing-cuthill-mckee/9268/11

**Comment by sbromberger** [comment]
Do we plan on including this in a separate submodule? What should we do with this PR?

**Comment by abhinavmehndiratta** [comment]
> Do we plan on including this in a separate submodule? What should we do with this PR?

There are no other reordering algorithms implemented at the moment but I could create a submodule if you feel it is needed.

**Comment by simonschoelly** [comment]
I think a separate Submodule would be overkill at this point. (Altough `squash!` would also be a candidate for preprocessing). I would suggest putting it to the Experimental submodule for know, then we are free to make changes later.

Also, a few days ago, someone on slack was asking for this functionality, so there is actually some demand.

**Pull request closed without merging**